rcl/src/rcl/qos_compatibility.c

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/qos_compatibility.h"
+
+#include "rcl/error_handling.h"
+#include "rcl/graph.h"
+#include "rcl/types.h"
+
+#include "rmw/qos_profiles.h"
+
+rcl_ret_t
+rcl_get_compatible_qos_for_topic_subscription(
+  const rcl_node_t * node,
+  const char * topic_name,
+  rmw_qos_profile_t * subscription_qos_profile)
+{
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(topic_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(subscription_qos_profile, RCL_RET_INVALID_ARGUMENT);
+
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  const rcl_allocator_t * allocator = &node_options->allocator;
+
+  rcl_topic_endpoint_info_array_t publishers_info =
+    rcl_get_zero_initialized_topic_endpoint_info_array();
+  rcl_ret_t ret = rcl_get_publishers_info_by_topic(
+    node,
+    // TODO(jacobperron): Update rcl_get_publishers_info_by_topic API to take an rcl_allocator_t 
+    (rcutils_allocator_t *)allocator,
+    topic_name,
+    false,
+    &publishers_info);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  const size_t number_of_publishers = publishers_info.size;
+  if (0u == number_of_publishers) {
+    return RCL_RET_OK;
+  }
+
+  // Copy QoS profiles to array
+  rmw_qos_profile_t * publisher_qos_profiles = (rmw_qos_profile_t *)allocator->allocate(
+    number_of_publishers * sizeof(rmw_qos_profile_t),
+    allocator->state);
+  for (size_t i = 0u; i < number_of_publishers; ++i) {
+    publisher_qos_profiles[i] = publishers_info.info_array[i].qos_profile;
+  }
+
+  rmw_ret_t rmw_ret = rmw_qos_profile_get_most_compatible_for_subscription(
+    publisher_qos_profiles,
+    number_of_publishers,
+    subscription_qos_profile,
+    NULL);
```

Comment:
Question: what should we do with this parameter, i.e. if one or more endpoints are not compatible?

I'm thinking this function could have an optional output parameter that is set to endpoint info... but I'm not sure what higher-level tooling will do with that information.

Reply:
> Question: what should we do with this parameter, i.e. if one or more endpoints are not compatible?

Return an error code.
The proposed rmw API also seems complex, I would just return an error if finding one is not possible, and not have an array of bools.

Reply:
Do we need a rmw function?
Can we have just rules of what's supposed to happen?
IMO, it would be weird to have different rules for different rmw.

Reply:
My thinking was that rules for QoS compatibility could be different for non-DDS rmw implementations. But since I don't have any concrete examples, it might be better at this point in time to only have an function in rcl and revisit an rmw function if/when we need it.

> I would just return an error if finding one is not possible, and not have an array of bools.

Yes, if we assume DDS rules, then I think we can remove this parameter since it is always possible to find a matching QoS. Even if there are "unknown" policies in the input, we can always return policies that will match all possible publishers.

Reply:
We discussed this in the ROS 2 meeting. My feeling was that we should keep the rmw function so that we can update the behavior, if needed, for future non-DDS implementations.